source/local/local.go

```diff
@@ -133,6 +168,8 @@ func parseFeatures(lines [][]byte) map[string]string {
 
 // Run all hooks and get features
 func getFeaturesFromHooks() (map[string]string, error) {
+	klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
```

Comment:
If this is logged every time hooks are called (which happens periodically), it could be too much.

IMHO it would be better to log the warning only at start, when reading local config, if HooksEnabled is set.

@marquiz?

(Otherwise looks OK to me.)

Reply:
I agree on this point. Initially I added in https://github.com/kubernetes-sigs/node-feature-discovery/blob/3df81d1ace1f74b43a74f998f8b238df6a53ca6a/pkg/nfd-client/worker/nfd-worker.go#L156 but then moved to here. Would https://github.com/kubernetes-sigs/node-feature-discovery/blob/3df81d1ace1f74b43a74f998f8b238df6a53ca6a/pkg/nfd-client/worker/nfd-worker.go#L156 be okay ?

Reply:
it would basically log once, as general warning regardless user enables/disabled the hooks.

Reply:
@marquiz should really be the one commenting on where to do the check...

(I filed the ticket for hooks disabling + few other things, so Markus CC'd me to this. However, this is first time I'm looking at NFD code, so I'm not at all familiar how it works.)

Reply:
Thanks @eero-t for the review.

@fmuyassarov I would probably put the log message later in this function if some hook files are really found and executed 🤔 Better not to print a warning for people who don't use hooks (most of them, I think). We could ratelimit into only one message e.g. with a boolean flag (that could be reseted e.g. in `SetConfig()` of the local source.

Reply:
Sure thing @marquiz . So we have two options I guess...
**1.** in `SetConfig()`, 

```
// SetConfig method of the LabelSource interface
func (s *localSource) SetConfig(conf source.Config) {
	switch v := conf.(type) {
	case *Config:
		s.config = v
		if s.config.HooksEnabled {
			klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
		}
	default:
		klog.Fatalf("invalid config type: %T", conf)
	}
}
```
log trace when hooks enabled
```
I0826 07:54:40.275016       1 nfd-worker.go:154] Node Feature Discovery Worker v0.12.0-devel-76-ga3100ce-dirty
I0826 07:54:40.275237       1 nfd-worker.go:155] NodeName: 'nfd-worker'
I0826 07:54:40.275465       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
W0826 07:54:40.275535       1 local.go:80] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:54:40.275547       1 nfd-worker.go:460] worker (re-)configuration successfully completed
I0826 07:54:40.275592       1 base.go:127] connecting to nfd-master at nfd-master:8080 ...
I0826 07:54:40.275630       1 component.go:36] [core]parsed scheme: ""
I0826 07:54:40.275640       1 component.go:36] [core]scheme "" not registered, fallback to default scheme

```
log trace when hooks enabled
```
I0826 07:58:39.551162       1 nfd-worker.go:154] Node Feature Discovery Worker v0.12.0-devel-76-ga3100ce-dirty
I0826 07:58:39.551321       1 nfd-worker.go:155] NodeName: 'nfd-worker'
I0826 07:58:39.552185       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
I0826 07:58:39.552341       1 nfd-worker.go:460] worker (re-)configuration successfully completed
I0826 07:58:39.552453       1 base.go:127] connecting to nfd-master at nfd-master:8080 ...
I0826 07:58:39.552612       1 component.go:36] [core]parsed scheme: ""
I0826 07:58:39.552627       1 component.go:36] [core]scheme "" not registered, fallback to default scheme
```

**2.** in `getFeaturesFromHooks()`. When hooks are enabled, depreciation message gets printed on every reload
```
func getFeaturesFromHooks(s *localSource) (map[string]string, error) {
.....
	if s.config.HooksEnabled {
		klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
	}
}
```
log trace when hooks enabled
```

I0826 07:42:56.606694       1 local.go:116] starting hooks...
W0826 07:42:56.618914       1 local.go:203] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:42:56.645574       1 nfd-worker.go:471] starting feature discovery...
I0826 07:42:56.646019       1 nfd-worker.go:483] feature discovery completed
I0826 07:42:56.646139       1 nfd-worker.go:564] sending labeling request to nfd-master
E0826 07:43:56.700575       1 kernel.go:132] failed to read kconfig: failed to read kernel config from [ /proc/config.gz /host-usr/src/linux-5.15.0-46-generic/.config /host-usr/src/linux/.config /host-usr/lib/modules/5.15.0-46-generic/config /host-usr/lib/ostree-boot/config-5.15.0-46-generic /host-usr/lib/kernel/config-5.15.0-46-generic /host-usr/src/linux-headers-5.15.0-46-generic/.config /lib/modules/5.15.0-46-generic/build/.config /host-boot/config-5.15.0-46-generic]
I0826 07:43:56.702207       1 local.go:116] starting hooks...
W0826 07:43:56.709296       1 local.go:203] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:43:56.735445       1 nfd-worker.go:471] starting feature discovery...
I0826 07:43:56.735863       1 nfd-worker.go:483] feature discovery completed
I0826 07:43:56.735929       1 nfd-worker.go:564] sending labeling request to nfd-master
```
log trace when hooks disabled
```
I0826 07:44:05.110774       1 nfd-worker.go:211] reloading configuration
I0826 07:44:05.116216       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
I0826 07:44:05.116438       1 nfd-worker.go:460] worker (re-)configuration successfully completed
E0826 07:44:05.119196       1 kernel.go:132] failed to read kconfig: failed to read kernel config from [ /proc/config.gz /host-usr/src/linux-5.15.0-46-generic/.config /host-usr/src/linux/.config /host-usr/lib/modules/5.15.0-46-generic/config /host-usr/lib/ostree-boot/config-5.15.0-46-generic /host-usr/lib/kernel/config-5.15.0-46-generic /host-usr/src/linux-headers-5.15.0-46-generic/.config /lib/modules/5.15.0-46-generic/build/.config /host-boot/config-5.15.0-46-generic]
I0826 07:44:05.145666       1 nfd-worker.go:471] starting feature discovery...
I0826 07:44:05.146036       1 nfd-worker.go:483] feature discovery completed
I0826 07:44:05.146066       1 nfd-worker.go:564] sending labeling request to nfd-master
```

 I personally prefer option 1 but fine with other opinions too. WDYT?

Reply:
Perhaps it is better to keep it as it is since the logs get printed only when hooks are enabled? It is true that users might not be using it although enabled and still see the logs, but it seems fine to me because we warning them about the feature that they should not enable. WDYT @marquiz ?

Reply:
In my thinking this might create confusion if you get warning like this with the default configuration (and you're not using hooks). But maybe it's good to print anyway - but in this case I'd probably do some rate limiting as I suggested, e.g. once after every config change or smth. WDYT?